flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java

```diff
@@ -0,0 +1,252 @@
+
+                        LOG.info(
+                                "Task {} has no state updates since last materialization, "
+                                        + "skip this one and schedule the next one in {} seconds",
+                                subtaskName,
+                                periodicMaterializeDelay / 1000);
+                    }
+                },
+                "materialization");
+    }
+
+    private void asyncMaterializationPhase(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture,
+            SequenceNumber upTo) {
+
+        SnapshotResult<KeyedStateHandle> materializedSnapshot = null;
+        FileSystemSafetyNet.initializeSafetyNetForThread();
+        try {
+            FutureUtils.runIfNotDoneAndGet(materializedRunnableFuture);
+
+            LOG.debug("Task {} finishes asynchronous part of materialization.", subtaskName);
+
+            materializedSnapshot = materializedRunnableFuture.get();
+
+        } catch (Exception e) {
+            int retryTime = numberOfConsecutiveFailures.incrementAndGet();
+
+            LOG.info(
+                    "Task {} asynchronous part of materialization is not completed for the {} time.",
+                    subtaskName,
+                    retryTime,
+                    e);
+
+            handleExecutionException(materializedRunnableFuture);
+
+            if (retryTime >= allowedNumberOfFailures) {
+                // Fail the task externally
+                asyncExceptionHandler.handleAsyncException(
+                        "Task "
+                                + subtaskName
+                                + " fails to complete the asynchronous part of materialization",
+                        e);
+
+                return;
+            }
+        } finally {
+            FileSystemSafetyNet.closeSafetyNetAndGuardedResourcesForThread();
+        }
+
+        // if succeed, update state and finish up
+        if (materializedSnapshot != null) {
+
+            final SnapshotResult<KeyedStateHandle> copyMaterializedSnapshot = materializedSnapshot;
+
+            mailboxExecutor.execute(
+                    () -> {
+                        keyedStateBackend.updateChangelogSnapshotState(
+                                copyMaterializedSnapshot, upTo);
+                        numberOfConsecutiveFailures.set(0);
+                    },
+                    "Task {} update materializedSnapshot up to changelog sequence number: {}",
+                    subtaskName,
+                    upTo);
+        }
+
+        LOG.info(
+                "Task {} schedules the next materialization in {} seconds.",
+                subtaskName,
+                periodicMaterializeDelay / 1000);
+
+        scheduleNextMaterialization();
+    }
+
+    private void handleExecutionException(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture) {
```

Comment:
There is already error handling logic calling this method. How about renaming this method to `discardFailedUpload` or something like this ?